internal/controllers/common/resource_definitions/resource_definitions.go

```diff
@@ -993,14 +1054,33 @@ func NewGrafanaContainer(cr *operatorv1beta1.Cryostat, imageTag string, tls *TLS
 				},
 			},
 		},
-		Resources:       cr.Spec.Resources.GrafanaResources,
 		SecurityContext: containerSc,
+		Resources:       NewGrafanaContainerResource(cr),
 	}
 }
 
 // datasourceURL contains the fixed URL to jfr-datasource's web server
 var datasourceURL = "http://" + loopbackAddress + ":" + strconv.Itoa(int(datasourceContainerPort))
 
+func NewJfrDatasourceContainerResource(cr *operatorv1beta1.Cryostat) corev1.ResourceRequirements {
+	resources := corev1.ResourceRequirements{}
+	if cr.Spec.Resources != nil {
+		resources = cr.Spec.Resources.DataSourceResources
+	}
+	if resources.Requests == nil {
+		resources.Requests = corev1.ResourceList{}
+	}
+	requests := resources.Requests
+	if _, found := requests[corev1.ResourceCPU]; !found {
+		requests[corev1.ResourceCPU] = resource.MustParse(defaultJfrDatasourceCpuRequest)
+	}
+	if _, found := requests[corev1.ResourceMemory]; !found {
+		requests[corev1.ResourceMemory] = resource.MustParse(defaultJfrDatasourceMemoryRequest)
+	}
+	checkResourceRequestWithLimit(requests, resources.Limits)
+	return resources
```

Comment:
This block of code is repeated for each container. Is it possible to extract it to a helper function to reduce duplication?

Reply:
Righ, i will try to refactor this.